Sources/Xcodeproj/XcodeProjectModelSerialization.swift

```diff
@@ -0,0 +1,576 @@
+        resultDict[name] = value
+    }
+    return .dictionary(resultDict)
+}
+
+
+/// A simple property list serializer with the same semantics as the Xcode
+/// property list serializer.  Not generally reusable at this point, but only
+/// because of implementation details (architecturally it isn't tied to Xcode).
+fileprivate class PropertyListSerializer {
+    
+    /// Private struct that represents a strong reference to a serializable
+    /// object.  This prevents any temporary objects from being deallocated
+    /// during the serialization and replaced with other objects having the
+    /// same object identifier (a violation of our assumptions)
+    struct SerializedObjectRef: Hashable, Equatable {
+        let object: PropertyListSerializable
+        init(_ object: PropertyListSerializable) {
+            self.object = object
+        }
+        var hashValue: Int {
+            return ObjectIdentifier(object).hashValue
+        }
+        static func ==(lhs: SerializedObjectRef, rhs: SerializedObjectRef) -> Bool {
+            return lhs.object === rhs.object
+        }
+    }
+    
+    /// Maps objects to the identifiers that have been assigned to them.  The
+    /// next identifier to be assigned is always one greater than the number
+    /// of entries in the mapping.
+    var objsToIds = [SerializedObjectRef: String]()
+    
+    /// Maps serialized objects ids to dictionaries.  This may contain fewer
+    /// entries than `objsToIds`, since ids are assigned upon reference, but
+    /// plist dictionaries are created only upon actual serialization.  This
+    /// dictionary is what gets written to the property list.
+    var idsToDicts = [String: PropertyList]()
+    
+    /// Returns the identifier for the object, assigning one if needed.
+    func id(of object: PropertyListSerializable) -> String {
+        // FIXME: This logic is ugly and should be rewritten to be elegant.
+        let serObjRef = SerializedObjectRef(object)
+        if let id = objsToIds[serObjRef] {
+            return id
+        }
+        // We currently always assign identifiers starting at 1 and going up.
+        // FIXME: This is a suboptimal format for object identifier strings;
+        // for debugging purposes they should at least sort in numeric order.
+        let id = "OBJ_\(objsToIds.count + 1)"
+        
+        objsToIds[serObjRef] = id
+        return id
+    }
+    
+    /// Serializes `object` by asking it to construct a plist dictionary and
+    /// then adding that dictionary to the serializer.  This may in turn cause
+    /// recursive invocations of `serialize(object:)`; the closure of these
+    /// invocations end up serializing the whole object graph.
+    @discardableResult func serialize(object: PropertyListSerializable) -> String {
+        let id = self.id(of: object)
+        precondition(idsToDicts[id] == nil, "tried to serialize \(object) twice")
+        idsToDicts[id] = .dictionary(object.serialize(to: self))
```

Comment:
Should this write a sentinel into `idsToDicts[id]` before invoking serialize in order to catch recursion?